File: content/LLD/Reference-Code/14. Builder Design Pattern/bad_Example.py
class Laptop:
    def __init__(
        self,
        processor: str,
        ram: str,
        graphic_card: str = None,
        color: str = None,
        screen_size: str = None,
    ):
        self.processor = processor
        self.ram = ram
        self.graphic_card = graphic_card
        self.color = color
        self.screen_size = screen_size

    def display_specs(self):
        print(f"Processor = {self.processor}")
        print(f"Ram = {self.ram}GB")
        if self.graphic_card:
            print(f"Graphic Card = {self.graphic_card}")
        if self.color:
            print(f"Color = {self.color}")
        if self.screen_size:
            print(f"Screen Size = {self.screen_size}")


laptop1 = Laptop("i5-3232", "6")
laptop1.display_specs()

laptop2 = Laptop("i5-3232", "6", None, "Black", None)
laptop2.display_specs()

File: content/LLD/Reference-Code/14. Builder Design Pattern/BadExample.java
public class BadExample {
    static class Laptop {
        private String processor;
        private String ram;
        private String graphicCard;
        private String color;
        private String screenSize;

        public Laptop(String processor, String ram, String graphicCard, String color, String screenSize) {
            this.processor = processor;
            this.ram = ram;
            this.graphicCard = graphicCard;
            this.color = color;
            this.screenSize = screenSize;
        }

        public Laptop(String processor, String ram) {
            this(processor, ram, null, null, null);
        }

        public void displaySpecs() {
            System.out.println("Processor = " + this.processor);
            System.out.println("Ram = " + this.ram + "GB");
            if (this.graphicCard != null) {
                System.out.println("Graphic Card = " + this.graphicCard);
            }
            if (this.color != null) {
                System.out.println("Color = " + this.color);
            }
            if (this.screenSize != null) {
                System.out.println("Screen Size = " + this.screenSize);
            }
        }
    }

    public static void main(String[] args) {
        Laptop laptop1 = new Laptop("i5-3232", "6");
        laptop1.displaySpecs();

        Laptop laptop2 = new Laptop("i5-3232", "6", null, "Black", null);
        laptop2.displaySpecs();
    }
}
